### DISCUSSION TITLE
how to secure RLS without auth

### DISCUSSION THREAD

**Discussion by pierroo** [supabase/supabase#20311] [Questions] [answered]
Hi,

So I have an old app with hundreds of thousands MAU;
It is still using firebase auth.

I am adding a new feature that will rely on supabase. (since the long term plan is to migrate to supabase eventually, I wanna test how it scales in production)

But the problem is: how can I secure access to those supabase tables etc, when most of RLS rely on auth and jwt?
I don't want to create yet again a new authentication in parallel of my existing firebase auth (also because supabase auth has a cost which woulds add hundreds of dollars to my monthly bill just with 250K+ MAU)

I guess it also depends on the logic I am trying to set up, the idea is to have a live match-making to have discussion with random live users.
So I have live_users table; chat_room & chat_msg

I shouldn't be able to view everyone who is online from front, only myself (based on a certain uuid generated from a previous rpc to add myself to pool of live_users)
and able to view/add messages in chat_msg depending on the chat_room_id foreign key for the room I belong to

I could pass from front the firebase uid and use it to limit through RLS, but I'm afraid it's unsafe since that firebase uid is visible in many other part of the app for anyone with a bit of networking / hacking skills so they could mess around supabase

Any help would be greatly appreciated

**Comment by GaryAustin1** [comment, marked as answer]
RLS with jwt's can be what ever you want it to be.  You just need to generate a user jwt from your firebase data with the same jwt secret key for both firebase and Supabase.  Or you can extract the info you want from firebase and generate a jwt with claims for supabase to use with a jwt library serverside.
RLS access the jwt with auth.jwt().  This can get to any claim in the jwt.  You would need to set the role claim to authenticated and if you set a sub claim to a UUID then you could use auth.uid().

**Reply by pierroo** [reply]
Wait that's super interesting I had no clue;
It's a bit over my current knowledge of JWT management so please bear with me:

- Obviously I have a jwt from firebase, but as far as I know we have no "jwt secret key" in firebase end; and even if we did (perhaps part of the config json file firebase provides?), how can I implement that same secret in supabase logic / dashboard, where would that go?
- the idea is to NOT have to rely on another middleware / server beyond the supabase dashboard here; so not sure how to connect supabase and firebase jwt here; or that jwt library serverside you mentioned, ideally I would want the whole jwt management to happen from RLS directly

So to summarize, how can I send the firebase token to supabase RLS for it to understand it through auth.uid()?
any documentation on this matter or is it too specific?

Thank you!

EDIT:
I found this, but this seems more web oriented (I'm trying to do it from a RN app)
https://liviogamassia.medium.com/using-supabase-rls-with-firebase-auth-custom-auth-provider-357eaad9c70f

and then found a comment in this thread https://github.com/supabase/gotrue-js/pull/340
from pixtron:
![image](https://github.com/supabase/supabase/assets/17619412/584c65e5-6c30-4104-b451-409f275b38fd)

is this the way to go somehow?

**Reply by GaryAustin1** [reply]
I just assumed there was an available secret for the Firebase JWTs.  If so you can set the Supabase secret to the same thing.  I've seen this done for other environments.

If not then you would need a middle layer server or edge function to take the Firebase jwt verify it and convert it into a Supabase jwt.
https://dev.to/sandrantsoele/firebase-auth-with-supabase-ka6 shows a sample of doing that.

Auth.uid() would depend on the jwt have a sub that is a UUID. But you don't need auth.uid() to use RLS.  You can use auth.jwt()->>'claim' to get anything out of the jwt and use that in RLS.

Related discussion and info:
https://github.com/orgs/supabase/discussions/391
docs on Supabase jwt: https://supabase.com/docs/learn/auth-deep-dive/auth-deep-dive-jwts#jwts-in-supabase
Mention of Firebase jwt secret:  https://answers.netlify.com/t/access-control-jwt-secret-in-netlify-dashboard-can-this-be-anything/79419/4

**Reply by pierroo** [reply]
Those links are great, I'm super close to have it done thanks to you.

My last question if you don't mind is (once again) related to pricing:
does creating a supabase token from a firebase one counts as auth pricing?

for example if I run that edge function to mint a supabase token for millions of users, and use it later through RLS, will it cost as per the supabase auth pricing?
Or it's completely unrelated?

**Reply by GaryAustin1** [reply]
The only auth "pricing" is the number of users signing in thru gotrue/auth per month which are counted as 1 access for a user no matter how many times that users signs in that month.    RLS and passing in a jwt do not count as auth as they are not going thru the gotrue REST API.  They will just be part of database egress costs.
